File: Problem Set 1 (code)/Cash.c
/**
 * Implemented without CS50 library
 * @author Stepan Ermakov
*/

#include <stdio.h>

int calculate_dimes(int change);
int calculate_nickels(int change);
int calculate_pennies(int change);
int calculate_quarters(int change);
int get_cents();

// A program that prompts the user for the number of cents 
// that a customer is owed and then prints the smallest number 
// of coins with which that change can be made
int main(void)
{
    int change = get_cents();

    int numberOfQuarters = calculate_quarters(change);
    change -= numberOfQuarters * 25;
    int numberOfDimes = calculate_dimes(change);
    change -= numberOfDimes * 10;
    int numberOfNickels = calculate_nickels(change);
    change -= numberOfNickels * 5;
    int numberOfPennies = calculate_pennies(change);

    int numberOfCoins = numberOfQuarters + numberOfDimes +
                        numberOfNickels + numberOfPennies;
    printf("%i\n", numberOfCoins);
}

int calculate_dimes(int change)
{
    int numberOfCoins = 0;
    int dime = 10;
    while (change >= dime)
    {
        numberOfCoins++;
        change -= dime;
    }
    return numberOfCoins;
}

int calculate_nickels(int change)
{
    int numberOfCoins = 0;
    int nickel = 5;
    while (change >= nickel)
    {
        numberOfCoins++;
        change -= nickel;
    }
    return numberOfCoins;
}

int calculate_pennies(int change)
{
    int numberOfCoins = 0;
    int penny = 1;
    while (change >= penny)
    {
        numberOfCoins++;
        change -= penny;
    }
    return numberOfCoins;
}

int calculate_quarters(int change)
{
    int numberOfCoins = 0;
    int quarter = 25;
    while (change >= quarter)
    {
        numberOfCoins++;
        change -= quarter;
    }
    return numberOfCoins;
}

int get_cents()
{
    int change;
    do
    {
        change = get_int("How much change is owed?\n");
    }
    while (change < 0);
    return change;
}

int get_int(char *string)
{
    
}

File: Problem Set 6/cash.py
# A program that prompts the user for the number of cents
# that a customer is owed and then prints the smallest number
# of coins with which that change can be made
__author__ = "Ermakov Stepan"


# Present [quarter, dime, nickel, penny]
coins = [25, 10, 5, 1]


def main():
    change = int(100 * get_float("Change owed: "))
    start = 0
    print(calculate_change(change, coins[start], start))


def calculate_change(change, coin, current):
    # Base-case
    if change == 0:
        return 0

    if change >= coin:
        return 1 + calculate_change(change - coin, coins[current], current)
    current += 1
    return 0 + calculate_change(change, coins[current], current)


def get_float(str):
    while True:
        try:
            value = float(input(str))
            if value > 0:
                break
        except ValueError:
            print("That's not a float!")
    return value


if __name__ == "__main__":
    main()

